### ISSUE TITLE
User guide / language documentation

### ISSUE THREAD

**Issue by jstolarek** [links-lang/links#711]
We currently don't have any kind of User Guide/language documentation that would be maintained and up to date. I've been thinking we could start pretending that we have one. The idea is that we actually create a stub of a guide and every time we add a new feature to the language, change existing syntax, add new compiler setting, etc., we document it in the guide. We almost do this already: most of the features we implement get a fairly detailed description in PRs, tickets, or changelogs. There would be some extra effort needed to add this into the guide but it seems like we're already doing 80% of the documenting effort (which then sadly gets buried in old PRs that nobody reads). With time that should give us a fairly good documentation and perhaps then filling in the missing gaps won't be too hard.

Does that sound like a good idea? Would all the devs be willing to document new features they are adding?

As for the technical side I'm not sure what would be the best, but I'd be for some markup language. Maybe Sphinx? Seems fairly lightweight.

Related: #474

**Comment by SimonJF** [comment]
We do something like this with the Wiki for new features: "https://github.com/links-lang/links/wiki"

But I agree, a more comprehensive getting started guide would be good. I think maybe taking `quick-help` and updating / expanding it and making it a proper web page would be a good idea.

**Comment by SimonJF** [comment]
I'd certainly be willing to help -- it would save time in the long run when onboarding new researchers / students / interns.

**Comment by jstolarek** [comment]
> We do something like this with the Wiki for new features:

Ah yes, I forgot to mention the wiki. My view is that wiki should be for the developers and guide for the users.

> making it a proper web page 

I think that Sphinx would allow us to compile both in PDF and HTML. Also, when we're discussing making a guide into a web page - which, I think is a Good Idea - I'd like to make sure that we can still work on some form of markup, rather than having to edit web pages.

Also, a downside of Sphinx is that it uses rst markup and not Markdown, which makes it difficult to copy what we already have on github.

**Comment by jamescheney** [comment]
Yes, this was partly the idea of asking people to write / update wiki pages, which I've been trying to normalize but doesn't quite happen regularly.  I agree that it would be better to migrate the ones that we want normal people to read to a clearly-advertised documentation system.  Ideally, one that supports multiple versions, so that when we create a new version of Links we can update the install instructions and any other parts without losing the old versions.  (In fact, this also seems like a good use case for archiving and provenance tracking.)

That said, weirdly students often seem to be able to get going using quick-help.html even though it has barely been updated for 10 years or so.  (Perhaps hiding all of Links's advanced features is a feature, not a bug, from the point of view of a new user, like with Racket's language levels.)

Somewhere there is a simple Links wiki and markdown parser (written using parser combinators), borrowed originally from SEWiki, a wiki with security labels that Nik Swamy and others wrote using SELinks.  I had in mind to resurrect this and adapt it into a documentation system, as a form of dogfooding.  We could host it on the VPS.  If anyone wants to pick that up, I can dig up the code and put it into a repository somewhere.

Alternatively, based on previous experience, this might be within the powers of a MSc or UG4 project student, and I'm down with using something standard for now and farming out the Links-dogfooding part to a future student.

**Comment by jamescheney** [comment]
We should also write an O'Reilly book called "Web Programming with Links".  For this we would probably want to be able to convert whatever markdown we use to DocBook.  I'm only 3/4 joking.

**Comment by jstolarek** [comment]
The problem with updating the wiki pages is that it cannot be reviewed as part of a PR and thus it often gets neglected, as you rightly pointed out.

> Ideally, one that supports multiple versions, so that when we create a new version of Links we can update the install instructions and any other parts without losing the old versions

What do you mean by that? I would expect that the user guide always describes the current version of Links. If we make user guide a part of the release then archival versions will be accessible just in the same way older versions of Links are.

> I'm down with using something standard for now and farming out the Links-dogfooding part to a future student.

Yes. The idea of using the said markdown parser etc. seems like a recipe for postponing things indefinitely.

> We should also write an O'Reilly book called "Web Programming with Links".

We might. Given that [there is already a book on an exotic framework like Yesod](http://shop.oreilly.com/product/0636920023142.do) the idea does not seem so weird.

**Comment by jamescheney** [comment]
> What do you mean by that? I would expect that the user guide always describes the current version of Links. If we make user guide a part of the release then archival versions will be accessible just in the same way older versions of Links are.

Sorry, I should have been a little clearer: something like this

https://docs.python.org/3.7/tutorial/index.html

note that there is a dropdown menu at the top allowing to navigate to the earlier versions (but not, for example, to compare different versions or see who has changed what...)  This is maybe more important for a language like Python where nontrivial numbers of people are using both 2.x and 3.x.  For a "user guide" that comes with the distribution maybe not so important as you say, or farmable out to a student project at least.

> Yes. The idea of using the said markdown parser etc. seems like a recipe for postponing things indefinitely.

OK

> Given that there is already a book on an exotic framework like Yesod the idea does not seem so weird.

Well, I think the number of developers using Yesod must be in the triple digits by now.

**Comment by SimonJF** [comment]
I have begun getting the Sphinx scaffolding in place, and begin scoping out what needs to happen.

Documentation development will happen here: https://github.com/links-lang/documentation

My current plan is to repurpose the `INSTALL.md`, `quick-help.pod`, and some of the Wiki pages and fill out various parts of the documentation from that. After, I'll set about filling in the blanks, documenting the Prelude, and then start on the more advanced features.

**Comment by jstolarek** [comment]
@SimonJF, thanks for doing this! There is just one thing that I disagree on, and I must say it is a very string opinion: having documentation in a separate repo. Documentation should ship with the project it concerns. In particular, with this kind of setup it will not be possible to reviews documentation as part of a PR and we will be in the same situation as with the wiki that technically exists but practically no one remembers about it because it is separated in a different repo. Working on two separate repos to update the documentation will be very inconvenient. So I insist that we have a user guide in the `doc` directory.

**Comment by SimonJF** [comment]
I agree, that's a good plan. I think my rationale for now is in getting the documentation up to scratch -- there's still a lot to do, and I'd rather just be able to commit and chip away at it, rather than making a deluge of PRs.

So my plan would be to get the basics filled out on the `documentation` repository, then port it over. Would that work?

**Comment by jstolarek** [comment]
> So my plan would be to get the basics filled out on the documentation repository, then port it over. Would that work?

If you feel this works best for you then sure, why not. Personally I would choose to work on a branch - that's what they are for.

**Comment by jstolarek** [comment]
I mean, there's no need to turn a branch into a PR until it's finished.

**Comment by SimonJF** [comment]
I think you're right -- I find myself agreeing with your proposal and am just being lazy in not converting it to a branch now that it's a separate repository :)

**Comment by jamescheney** [comment]
Now that #737 is merged, I suggest to close this and create new issues in the "Documentation..." project on specific things that need to be documented.
